From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Wed, 19 Aug 2020 09:17:05 +0300
Subject: [PATCH] [sailfishos][gecko] Hackish fix for preferences usage in
 Parent process (part 1)

Inspired by SHA1: 18b3279b0851758df8805b760e9026d45fcf142c Mon Sep 17 00:00:00 2001
From: Tatiana Meshkova <tanya.meshkova@gmail.com>
Date: Tue, 20 Jan 2015 23:49:38 -0800

Hackish fix for preferences usage in Parent process. Added basic
compositor initialization in Parent process

This applies patch 0021 from ESR 78.

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
Signed-off-by: Pavel Tumakaev <p.tumakaev@omprussia.ru>
---
 dom/ipc/DOMTypes.ipdlh                         | 3 +++
 gfx/layers/composite/LayerManagerComposite.cpp | 4 +++-
 gfx/layers/ipc/LayerTransactionParent.h        | 5 +++++
 gfx/thebes/gfxPlatform.cpp                     | 2 +-
 modules/libpref/Preferences.cpp                | 7 +++++++
 5 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/dom/ipc/DOMTypes.ipdlh b/dom/ipc/DOMTypes.ipdlh
index 38efbe357823..fdb1d2f158ca 100644
--- a/dom/ipc/DOMTypes.ipdlh
+++ b/dom/ipc/DOMTypes.ipdlh
@@ -27,6 +27,9 @@ using struct mozilla::void_t from "mozilla/ipc/IPCCore.h";
 using class mozilla::dom::LoadingSessionHistoryInfo
   from "mozilla/dom/SessionHistoryEntry.h";
 
+using struct mozilla::null_t
+  from "ipc/IPCMessageUtils.h";
+
 using LayoutDeviceIntRect from "Units.h";
 using DesktopIntRect from "Units.h";
 using DesktopToLayoutDeviceScale from "Units.h";
diff --git a/gfx/layers/composite/LayerManagerComposite.cpp b/gfx/layers/composite/LayerManagerComposite.cpp
index 11c1c06cc56d..72d1c487a64a 100644
--- a/gfx/layers/composite/LayerManagerComposite.cpp
+++ b/gfx/layers/composite/LayerManagerComposite.cpp
@@ -181,7 +181,9 @@ LayerManagerComposite::~LayerManagerComposite() { Destroy(); }
 
 void LayerManagerComposite::Destroy() {
   if (!mDestroyed) {
-    mCompositor->GetWidget()->CleanupWindowEffects();
+    if (mCompositor->GetWidget()) {
+      mCompositor->GetWidget()->CleanupWindowEffects();
+    }
     if (mRoot) {
       RootLayer()->Destroy();
     }
diff --git a/gfx/layers/ipc/LayerTransactionParent.h b/gfx/layers/ipc/LayerTransactionParent.h
index 417236729ae3..e275586819f0 100644
--- a/gfx/layers/ipc/LayerTransactionParent.h
+++ b/gfx/layers/ipc/LayerTransactionParent.h
@@ -18,6 +18,10 @@
 
 namespace mozilla {
 
+namespace embedlite {
+class EmbedLiteCompositorProcessParent;
+}
+
 namespace ipc {
 class Shmem;
 }  // namespace ipc
@@ -163,6 +167,7 @@ class LayerTransactionParent final : public PLayerTransactionParent,
   }
   friend class CompositorBridgeParent;
   friend class ContentCompositorBridgeParent;
+  friend class mozilla::embedlite::EmbedLiteCompositorProcessParent;
 
  private:
   // This is a function so we can log or breakpoint on why hit
diff --git a/gfx/thebes/gfxPlatform.cpp b/gfx/thebes/gfxPlatform.cpp
index 49a346d405fe..39460f1e5abd 100644
--- a/gfx/thebes/gfxPlatform.cpp
+++ b/gfx/thebes/gfxPlatform.cpp
@@ -1022,7 +1022,7 @@ void gfxPlatform::Init() {
   // Request the imgITools service, implicitly initializing ImageLib.
   nsCOMPtr<imgITools> imgTools = do_GetService("@mozilla.org/image/tools;1");
   if (!imgTools) {
-    MOZ_CRASH("Could not initialize ImageLib");
+//    MOZ_CRASH("Could not initialize ImageLib");
   }
 
   RegisterStrongMemoryReporter(new GfxMemoryImageReporter());
diff --git a/modules/libpref/Preferences.cpp b/modules/libpref/Preferences.cpp
index 6d62eeea11e0..d7e8b6d4ec13 100644
--- a/modules/libpref/Preferences.cpp
+++ b/modules/libpref/Preferences.cpp
@@ -90,6 +90,8 @@
 #  include "mozilla/BackgroundTasks.h"
 #endif
 
+#include "mozilla/embedlite/EmbedLiteAppProcessParent.h"
+
 #ifdef DEBUG
 #  include <map>
 #endif
@@ -3517,6 +3519,11 @@ Preferences::~Preferences() {
   PrefNameArena().Clear();
 }
 
+static mozilla::embedlite::EmbedLiteAppProcessParent* GetEmbedLiteParent()
+{
+  return mozilla::embedlite::EmbedLiteAppProcessParent::GetInstance();
+}
+
 NS_IMPL_ISUPPORTS(Preferences, nsIPrefService, nsIObserver, nsIPrefBranch,
                   nsISupportsWeakReference)
 
